src/herder/ParallelTxSetBuilder.cpp

```diff
@@ -434,87 +580,60 @@ buildSurgePricedParallelSorobanPhase(
-        bool added = false;
-        auto builderTxIt = builderTxForTx.find(tx);
-        releaseAssert(builderTxIt != builderTxForTx.end());
-        for (auto& stage : stages)
-        {
-            if (stage.tryAdd(*builderTxIt->second))
-            {
-                added = true;
-                break;
-            }
-        }
-        if (added)
-        {
-            return SurgePricingPriorityQueue::VisitTxResult::PROCESSED;
-        }
-        // If a transaction didn't fit into any of the stages, we consider it
-        // to have been excluded due to resource limits and thus notify the
-        // surge pricing queue that surge pricing should be triggered (
-        // REJECTED imitates the behavior for exceeding the resource limit
-        // within the queue itself).
-        return SurgePricingPriorityQueue::VisitTxResult::REJECTED;
-    };
+    // Create a worker thread for each stage count.
+    std::vector<std::thread> threads;
+    uint32_t stageCountOptions = cfg.SOROBAN_PHASE_MAX_STAGE_COUNT -
+                                 cfg.SOROBAN_PHASE_MIN_STAGE_COUNT + 1;
+    std::vector<ParallelPhaseBuildResult> results(stageCountOptions);
 
-    std::vector<Resource> laneLeftUntilLimitUnused;
-    queue.popTopTxs(/* allowGaps */ true, visitor, laneLeftUntilLimitUnused,
-                    hadTxNotFittingLane);
-    // There is only a single fee lane for Soroban, so there is only a single
-    // flag that indicates whether there was a transaction that didn't fit into
-    // lane (and thus all transactions are surge priced at once).
-    releaseAssert(hadTxNotFittingLane.size() == 1);
-
-    // At this point the stages have been filled with transactions and we just
-    // need to place the full transactions into the respective stages/clusters.
-    TxStageFrameList resStages;
-    resStages.reserve(stages.size());
-    for (auto const& stage : stages)
+    for (uint32_t stageCount = cfg.SOROBAN_PHASE_MIN_STAGE_COUNT;
+         stageCount <= cfg.SOROBAN_PHASE_MAX_STAGE_COUNT; ++stageCount)
     {
-        auto& resStage = resStages.emplace_back();
-        resStage.reserve(partitionCfg.mClustersPerStage);
-
-        std::unordered_map<size_t, size_t> clusterIdToStageCluster;
-
-        stage.visitAllTransactions([&resStage, &txFrames,
-                                    &clusterIdToStageCluster](size_t clusterId,
-                                                              size_t txId) {
-            auto it = clusterIdToStageCluster.find(clusterId);
-            if (it == clusterIdToStageCluster.end())
-            {
-                it = clusterIdToStageCluster.emplace(clusterId, resStage.size())
-                         .first;
-                resStage.emplace_back();
-            }
-            resStage[it->second].push_back(txFrames[txId]);
+        size_t resultIndex = stageCount - cfg.SOROBAN_PHASE_MIN_STAGE_COUNT;
+        threads.emplace_back([queue, &builderTxForTx, txFrames, stageCount,
```

Comment:
Are there any performance concerns here? I'm wondering if it'd make sense to set `SOROBAN_PHASE_MAX_STAGE_COUNT` to 3 instead of 4 if we don't expect to need 4 stages that often.

Reply:
3 is not good enough either; I think 1 is preferable when we can pull it off, as it allows for better ledger space utilization, as well as less apply-time friction. I'm not concerned about performance at the moment - we shouldn't be nominating tx sets while applying tx sets, so there should be enough capacity to spin up a few threads.